strawberry/channels/testing.py

```diff
@@ -125,22 +125,24 @@ async def subscribe(
             message_type = response["type"]
             if message_type == NextMessage.type:
                 payload = NextMessage(**response).payload
-                ret = ExecutionResult(None, None)
-                for field in dataclasses.fields(ExecutionResult):
-                    setattr(ret, field.name, payload.get(field.name, None))
-                    yield ret
+                ret = ExecutionResult(payload["data"], None)
+                if "errors" in payload:
+                    ret.errors = process_errors(payload["errors"])
+                ret.extensions = payload.get("extensions", None)
+                yield ret
             elif message_type == ErrorMessage.type:
                 error_payload = ErrorMessage(**response).payload
-                yield ExecutionResult(
-                    data=None,
-                    errors=[
-                        GraphQLError(
-                            message=message["message"],
-                            extensions=message.get("extensions", None),
-                        )
-                        for message in error_payload
-                    ],
-                )
+                yield ExecutionResult(data=None, errors=process_errors(error_payload))
                 return  # an error message is the last message for a subscription
             else:
                 return
+
+
+def process_errors(errors: List[GraphQLFormattedError]) -> List[GraphQLError]:
+    return [
+        GraphQLError(
+            message=error["message"],
+            extensions=error.get("extensions", None),
+        )
```

Comment:
Not specific to this change since it has been like this before, but I notice we transform errors differently in other places in our code. For example:

https://github.com/strawberry-graphql/strawberry/blob/8f18beba507e56973a555967ca584e3b3db2ae36/strawberry/schema/execute.py#L109

https://github.com/strawberry-graphql/strawberry/blob/8f18beba507e56973a555967ca584e3b3db2ae36/strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py#L314

Looking at the graphql-core source we can see this is also preserving extensions:

https://github.com/graphql-python/graphql-core/blob/0c93b8452eed38d4f800c7e71cf6f3f3758cd1c6/src/graphql/error/graphql_error.py#L169-L173

We might want to use this approach here as well :)


```suggestion
        GraphQLError(str(error), original_error=error)
```

Reply:
Hm, this change breaks typing, I'll investigate.

Reply:
Ah, this is a different case.  The code in the channels testing client is transforming the error for the _client_, whereas that other code you saw is creating GraphQLErrors on the server side.

I am frankly unsure why the channels testing client is doing this, since the error is already a GraphQLFormattedError.   Except as some sort of convenience for the testing framework.   I think this transformation should be dropped, in fact.

Reply:
Oh, the reason is that it is yielding an "ExecutionResult".  
The thing is, ExecutionResult is specified as containing "GraphQLError".  
But this is weird, because, on the wire, it really should be GraphQLFormattedError.

I think this stems from a confusion between python types and graphql types.  You cannot in fact create an ExecutionResult from the NextMessage.  An ExecutionResult is a server internal format, whereas the NextMessage contains formatted data, including formatted errors...

It should be a `FormattedExecutionResult` type...

Reply:
Oh, interesting, I see. Thanks for the investigation and write up! We can adress this in another PR. This one is good to go now.